File: training_data/reflection-tuning/day17_part1_2023.py
import heapq
from typing import List, Tuple

# Constants for directions
UP, RIGHT, DOWN, LEFT = (-1, 0), (0, 1), (1, 0), (0, -1)
DIRECTIONS = [UP, RIGHT, DOWN, LEFT]

def min_heat_loss(grid: List[List[int]]) -> int:
    rows, cols = len(grid), len(grid[0])
    target = (rows - 1, cols - 1)
    
    # Priority queue: (heat_loss, row, col, direction, steps_in_direction)
    pq = [(0, 0, 0, None, 0)]
    visited = set()
    
    while pq:
        heat_loss, row, col, direction, steps = heapq.heappop(pq)
        
        if (row, col) == target:
            return heat_loss
        
        if (row, col, direction, steps) in visited:
            continue
        visited.add((row, col, direction, steps))
        
        for new_dir in DIRECTIONS:
            if new_dir == direction and steps == 3:
                continue
            if direction and new_dir == (-direction[0], -direction[1]):
                continue
            
            new_row, new_col = row + new_dir[0], col + new_dir[1]
            if 0 <= new_row < rows and 0 <= new_col < cols:
                new_steps = 1 if new_dir != direction else steps + 1
                new_heat_loss = heat_loss + grid[new_row][new_col]
                heapq.heappush(pq, (new_heat_loss, new_row, new_col, new_dir, new_steps))
    
    return -1  # No path found

# Read input from file
with open('input.txt', 'r') as file:
    grid = [[int(char) for char in line.strip()] for line in file]

# Solve and print the result
print(min_heat_loss(grid))

File: rust/day17_part1_2023.rs

use std::{
    cmp::Ordering,
    collections::{BinaryHeap, HashMap},
    fs::File,
    io::{self, BufRead},
};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
enum Direction {
    Up,
    Down,
    Left,
    Right,
}

impl Direction {
    fn opposite(&self) -> Self {
        match self {
            Direction::Up => Direction::Down,
            Direction::Down => Direction::Up,
            Direction::Left => Direction::Right,
            Direction::Right => Direction::Left,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct State {
    row: usize,
    col: usize,
    dir: Direction,
    steps_in_dir: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct Node {
    state: State,
    cost: usize,
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> Ordering {
        other.cost.cmp(&self.cost)
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn solve(grid: &Vec<Vec<usize>>) -> usize {
    let rows = grid.len();
    let cols = grid[0].len();

    let start_state_right = State {
        row: 0,
        col: 0,
        dir: Direction::Right,
        steps_in_dir: 0,
    };

    let start_state_down = State {
        row: 0,
        col: 0,
        dir: Direction::Down,
        steps_in_dir: 0,
    };

    let mut dist: HashMap<State, usize> = HashMap::new();
    let mut pq: BinaryHeap<Node> = BinaryHeap::new();

    dist.insert(start_state_right, 0);
    dist.insert(start_state_down, 0);

    pq.push(Node {
        state: start_state_right,
        cost: 0,
    });

     pq.push(Node {
        state: start_state_down,
        cost: 0,
    });
   

    while let Some(Node { state, cost }) = pq.pop() {
        if state.row == rows - 1 && state.col == cols - 1 {
            return cost;
        }

        if cost > *dist.get(&state).unwrap_or(&usize::MAX) {
            continue;
        }

        let directions = [
             (Direction::Up, (-1, 0)),
             (Direction::Down, (1, 0)),
             (Direction::Left, (0, -1)),
             (Direction::Right, (0, 1)),
        ];

        for (next_dir, (dr, dc)) in directions.iter() {
           if next_dir == &state.dir.opposite() {
                continue;
           }
            let next_steps_in_dir = if *next_dir == state.dir {state.steps_in_dir+1} else {1};
            if next_steps_in_dir > 3 {
                continue;
            }

            let next_row = state.row as i32 + dr;
            let next_col = state.col as i32 + dc;

            if next_row >= 0
                && next_row < rows as i32
                && next_col >= 0
                && next_col < cols as i32
            {
                let next_row = next_row as usize;
                let next_col = next_col as usize;
                let next_state = State {
                    row: next_row,
                    col: next_col,
                    dir: *next_dir,
                    steps_in_dir: next_steps_in_dir,
                };
                let next_cost = cost + grid[next_row][next_col];
                if next_cost < *dist.get(&next_state).unwrap_or(&usize::MAX) {
                    dist.insert(next_state, next_cost);
                    pq.push(Node {
                        state: next_state,
                        cost: next_cost,
                    });
                }
            }
        }
    }
    unreachable!("No path found")
}

fn main() -> io::Result<()> {
    let file = File::open("input.txt")?;
    let reader = io::BufReader::new(file);

    let grid: Vec<Vec<usize>> = reader
        .lines()
        .map(|line| {
            line.unwrap()
                .chars()
                .map(|c| c.to_digit(10).unwrap() as usize)
                .collect()
        })
        .collect();

    let result = solve(&grid);
    println!("{}", result);

    Ok(())
}
